### ISSUE TITLE
uninstall unused runtimes (and extensions)

### ISSUE THREAD

**Issue by ramcq** [flatpak/flatpak#1747]
As well as processing autodeletes, Debug and Locales upon uninstall, Flatpak should also contain the ability to walk the metadata "down" from a certain app to do a reachability check of every extension, runtime, and runtime extension. (Possibly allowing a pass for SDKs...?)

It should be run in a targeted way when uninstalling an app so that when you remove the last app that uses a certain runtime or extension, those runtimes are also removed or offered to be removed, to free disk space.

It could also help implement a fix for #1746 to check whether any deletion candidate is in fact still in use by another flatpak.

A full reachability check might also make sense to include in the flatpak repair operation.

**Comment by matthiasclasen** [comment]
Does "flatpak uninstall --unused" in 0.11.8 cover this ?

https://github.com/flatpak/flatpak/releases

**Comment by ramcq** [comment]
Yes - although it would be nice to also have this in the API so that G-S can use it?

**Comment by alex285** [comment]
hello @matthiasclasen 
I'm not sure `flatpak uninstall --unused` works correctly though. For instance

```
flatpak remove --unused

Uninstalling from user:
 runtime/org.freedesktop.Platform.GL.nvidia-396-24/x86_64/1.4
 runtime/org.freedesktop.Platform.Icontheme.Adwaita/x86_64/1.0
 runtime/org.freedesktop.Platform.ffmpeg/x86_64/1.6
 runtime/org.gnome.Sdk.Docs/x86_64/3.28
 runtime/org.gnome.Sdk.Locale/x86_64/3.28
 runtime/org.gnome.Sdk/x86_64/3.28
 runtime/org.gtk.Gtk3theme.Adapta-Nokto/x86_64/3.22
 runtime/org.gtk.Gtk3theme.elementary/x86_64/3.22
Uninstalling from system:
 runtime/org.gimp.GIMP.Locale/x86_64/master
 runtime/org.gimp.GIMP.Locale/x86_64/stable
 runtime/org.kde.KStyle.Adwaita/x86_64/master
 runtime/org.mozilla.FirefoxNightly.Locale/x86_64/master
Is this ok? [y/n]: 
```

What does in this case is that also removes the themes just because are not currently used? And most importantly it even marks for removal ` runtime/org.gnome.Sdk/x86_64/3.28`

**Comment by alex285** [comment]
Actually it may works correctly, but for sure it is not a command line that someone can use to clean his system. It is more like "some suggestions" of  runtimes / extensions that we may diont need, and then remove them one by one

**Comment by alex285** [comment]
I think there is something terribly goings on here. I see themes and NVIDIA on `user`, but I never used a `--user` flag to install them, and I got them on `--system` too anyway

**Comment by matthiasclasen** [comment]
I think we can close this. The basic --unused functionality is in place.

**Comment by ramcq** [comment]
But it's not available in the API...?

**Comment by matthiasclasen** [comment]
I guess maybe the code that finds all 'used' refs in a dir could be made library api
